### PR TITLE
Fix exceptions

### PR DESCRIPTION
**Pull request by DemiMarie**
This patchset fixes exception handling in RustJni.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,8 +1,10 @@
-main
+# Emacs backup files
 *~
 .#*
 \#*\#
-*.rlib
```
Comment (Monnoroch) [inline .gitignore]: Why did you remove those?

**Comment by Monnoroch** [comment]
That's alot of code. I think it's time to write some tests. Will try to find some time for it.

**Comment by Monnoroch** [comment]
The problem with your implementation, that it, for example, forbids creating multiple JavaEnv's from a single JavaVM, which is totally wrong, at least for the multithreaded usage: there should be a JavaEnv attached for each native thread.
For now, i'll try to figure something out, but maby you have some thoughts on a better implementation?

**Comment by drbo** [comment]
You can call `get_env` more than once to obtain separate `JavaEnv`
objects.  There is a bug in that they do not return `Capability` objects
(my fault).

**Comment by Monnoroch** [comment]
Well, you also can't mutable borrow jvm reference several times, which apparently JavaEnv does.

**Comment by Monnoroch** [comment]
@drbo Hi again!
I've been thinking for a few days now: what exactly is the problem with exceptions, that you had to make this interesting, but rather complicated api?
I mean, i've tested some JNI calls, they work perfectly even there is an exception in flight.
Why the user must think about exceptions on each call? He can do it without this Capability/Exception objects, just by using standart exception api.

**Comment by drbo** [comment]
On Fri, 2015-05-29 at 08:32 -0700, Max wrote:
> @drbo Hi again!
> I've been thinking for a few days now: what exactly is the problem
> with exceptions, that you had to make this interesting, but rather
> complicated api?
> I mean, i've tested some JNI calls, they work perfectly even there is
> an exception in flight.
> Why the user must think about exceptions on each call? He can do it
> without this Capability/Exception objects, just by using standart
> exception api.
> 
> —
> Reply to this email directly or view it on GitHub.
> 
> 
According to the [Jni
documentation](http://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/design.html#exceptions_and_error_codes) JNI code is not permitted to call most JNI functions when a Java exception is pending.  This requirement is stated using "must", which means that violating it is undefined behavior as per the [JNI function specification](http://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html).
Therefore, all functions in RustJni must enforce this requirement unless
marked as `unsafe`.

**Comment by Monnoroch** [comment]
Yeah, thanks, I read that too.
